inventory/src/main/java/com/yas/inventory/service/StockService.java

```diff
@@ -50,4 +54,38 @@ public void addProductIntoWarehouse(List<StockPostVM> postVMs) {
 
     }
 
+    public void checkQuantityAvailable(List<StockCheckRequest> stockCheckRequests) {
+        List<Long> productIds = stockCheckRequests.stream().map(StockCheckRequest::productId).toList();
+        List<Stock> stocks = stockRepository.getAvailableStockByProductIds(productIds);
+        List<OutStockErrorVm> errors = new ArrayList<>();
+        stockCheckRequests.forEach(scr -> {
+            List<Stock> stockChecks = stocks.stream()
+                    .filter(stock -> scr.productId().equals(stock.getProductId())).toList();
+            if (!isValidOrderQuantity(stockChecks, scr.quantity())) {
+                errors.add(new OutStockErrorVm(scr.productId(), scr.quantity()));
+            }
+        });
```

Comment:
You can use a parallel stream instead of a sequential stream for the forEach loop that checks the StockCheckRequest objects against the available stock. This can potentially improve performance on multi-core machines.
i don't know this really works but try this
Map<Long, Stock> stockMap = stocks.stream().collect(Collectors.toMap(Stock::getProductId, Function.identity()));
    List<OutStockErrorVm> errors = new ArrayList<>();
    stockCheckRequests.parallelStream().forEach(scr -> {
        Stock stock = stockMap.get(scr.productId());
        if (stock == null || !isValidOrderQuantity(stock, scr.quantity())) {
            errors.add(new OutStockErrorVm(scr.productId(), scr.quantity()));
        }
    });

-  **### Conlusion, use parallel Stream to optimize performance**